hash_map.h

```diff
@@ -0,0 +1,192 @@
+#pragma once
+#include <iostream>
+#include <vector>
+#include <list>
+#include <stdexcept>
+
+template<
+    class KeyType, 
+    class ValueType, 
+    class Hash = std::hash<KeyType>>
+class HashMap {
+private:
+    using size_type = size_t;
+    using list_type = std::list<std::pair<const KeyType, ValueType>>;
+    using table_type = std::vector<std::list<typename list_type::iterator>>;
+
+    Hash hasher_;
+
+    list_type list_of_pairs_;
+    table_type table_;
+
+    const size_type kMinSize = 16;
+    const double alpha_ = 2;
+    size_type elements_stored_;
+    size_type size_;
+
+    void rehash() {
+        size_ *= 2;
+        table_type temp_table(size_, { list_of_pairs_.end() });

```

Comment:
```c++
table_.clear();
table_resize(size_);
```
Should works without temp_table.

Also temp table miss naming rules: https://google.github.io/styleguide/cppguide.html#General_Naming_Rules